text/2020-00-00-error-handling.md

```diff
@@ -0,0 +1,297 @@
+
+Current definition (from
+[src/storage/txn/mod.rs](https://github.com/tikv/tikv/blob/dec3d008acb56ad0fb34eeeecca62aaa9091b994/src/storage/txn/mod.rs#L42-L94)
+):
+
+```rust
+quick_error! {
+    #[derive(Debug)]
+    pub enum ErrorInner {
+        Engine(err: crate::storage::kv::Error) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        Mvcc(err: crate::storage::mvcc::Error) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        ProtoBuf(err: protobuf::error::ProtobufError) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        Io(err: IoError) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        InvalidTxnTso {start_ts: TimeStamp, commit_ts: TimeStamp} {
+            description("Invalid transaction tso")
+            display("Invalid transaction tso with start_ts:{},commit_ts:{}",
+                        start_ts,
+                        commit_ts)
+        }
+        Other(err: Box<dyn error::Error + Sync + Send>) {
+            from()
+            cause(err.as_ref())
+            description(err.description())
+            display("{:?}", err)
+        }
+    }
+}
+```
+
+Definition with proposed changes:
+
+```rust
+#[derive(Debug, thiserror::Error)]
+#[error(unwrap)]
+pub enum Error {
+    #[error(transparent)]
+    Engine(#[from] crate::storage::kv::Error),
+
+    #[error(transparent)]
+    Mvcc(#[from] crate::storage::mvcc::Error),
+
+    #[error(transparent)]
+    ProtoBuf(#[from] Box<protobuf::error::ProtobufError>),
+
+    #[error(transparent)]
+    Io(#[from] IoError),
+
+    #[error("Invalid transaction tso with start_ts: \
+             {start_ts}, commit_ts: {commit_ts}")]
+    InvalidTxnTso {
+        start_ts: TimeStamp,
+        commit_ts: TimeStamp,
+    },
+
+    #[error(transparent)]
+    Other(#[from] AnyError),
+}
+```
+
+Note that I've elided some variants for concision.
+
+Current pattern (from
+[src/storage/errors.rs](https://github.com/tikv/tikv/blob/dec3d008acb56ad0fb34eeeecca62aaa9091b994/src/storage/errors.rs#L174-L180)
+):
+
+```rust
+    Err(Error(box ErrorInner::Engine(EngineError(box EngineErrorInner::Request(e)))))
+    | Err(Error(box ErrorInner::Txn(TxnError(box TxnErrorInner::Engine(EngineError(
+        box EngineErrorInner::Request(e),
+    ))))))
+    | Err(Error(box ErrorInner::Txn(TxnError(box TxnErrorInner::Mvcc(MvccError(
+        box MvccErrorInner::Engine(EngineError(box EngineErrorInner::Request(e))),
+    )))))) => ...,
+```
+
+Pattern with proposed changes:
+
+```rust
+    Err(Error::Engine(EngineError::Request(e)))
+    | Err(Error::Txn(TxnError::Engine(EngineError::Request(e))))
+    | Err(Error::Txn(TxnError::Mvcc(MvccError::Engine(
+        EngineError::Request(e),
+    )))) => ...,
+```
+
+Once the error migration is complete, we are guaranteed that the second and
```

Comment:
What is the "error migration"? How can it remove the second and the third match arm?

Reply:
"error migration" means converting all our error types to the system proposed in this RFC. Due to the `unwrap` stuff, there are no deeply nested errors and we only ever have to look one level deep.